extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,238 @@
+						return parent;
+					} else {
+						return new vscode.SelectionRange(adjustedRange, parent.parent);
+					}
+				} else {
+					return parent;
+				}
+			} else {
+				return new vscode.SelectionRange(range);
+			}
+		}
+	} else {
+		return undefined;
+	}
+}
+
+function createFencedRange(token: Token, cursorLine: number, document: vscode.TextDocument, parent?: vscode.SelectionRange): vscode.SelectionRange {
+	const startLine = token.map[0];
+	const endLine = token.map[1] - 1;
+	let onFenceLine = cursorLine === startLine || cursorLine === endLine;
+	let fenceRange = new vscode.Range(new vscode.Position(startLine, 0), new vscode.Position(endLine, document.lineAt(endLine).text.length));
+	let contentRange = endLine - startLine > 2 && !onFenceLine ? new vscode.Range(new vscode.Position(startLine + 1, 0), new vscode.Position(endLine - 1, getEndCharacter(document, startLine + 1, endLine))) : undefined;
+	if (parent && contentRange) {
+		if (parent.range.contains(fenceRange) && !parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange, parent));
+		} else if (parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(contentRange, parent);
+		} else if (rangeLinesEqual(fenceRange, parent.range)) {
+			let revisedRange = fenceRange.end.character > parent.range.end.character ? fenceRange : parent.range;
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(revisedRange, getRealParent(parent, revisedRange)));
+		} else if (parent.range.end.line === fenceRange.end.line) {
+			parent.range.end.translate(undefined, fenceRange.end.character);
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange, parent));
+		}
+	} else if (contentRange) {
+		return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange));
+	} else if (parent) {
+		if (parent.range.contains(fenceRange) && !parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(fenceRange, parent);
+		} else if (parent.range.isEqual(fenceRange)) {
+			return parent;
+		} else if (rangeLinesEqual(fenceRange, parent.range)) {
+			let revisedRange = fenceRange.end.character > parent.range.end.character ? fenceRange : parent.range;
+			return new vscode.SelectionRange(revisedRange, parent.parent);
+		} else if (parent.range.end.line === fenceRange.end.line) {
+			parent.range.end.translate(undefined, fenceRange.end.character);
+			return new vscode.SelectionRange(fenceRange, parent);
+		}
+	}
+	return new vscode.SelectionRange(fenceRange, parent);
+}
+
+function isList(type: string): boolean {
+	return type ? ['ordered_list_open', 'list_item_open', 'bullet_list_open'].includes(type) : false;
+}
+
+function getEndCharacter(document: vscode.TextDocument, startLine: number, endLine: number): number {
+	let startLength = document.lineAt(startLine).text ? document.lineAt(startLine).text.length : 0;
```

Comment:
[super nit] This will potentially call ` document.lineAt(startLine)` twice. You can use typescript null collapsing operators to simplify the code and avoid this:

```ts
// Access length if it exists, or return 0
document.lineAt(startLine).text?.length ?? 0
```